R/utils.R

```diff
@@ -92,3 +92,123 @@ browse_url <- function(path, chromote) {
     utils::browseURL(url)
   }
 }
+
+
+# =============================================================================
+# Random Ports
+# =============================================================================
+#
+# Borrowed from https://github.com/rstudio/httpuv/blob/main/R/random_port.R
+
+with_random_port <- function(
+  startup,
+  min = 1024L,
+  max = 49151L,
+  n = 5,
+  host = "http://127.0.0.1"
+) {
+  stopifnot(is.function(startup))
+  valid_ports <- setdiff(seq.int(min, max), unsafe_ports)
+
+  # Try up to n ports
+  n <- min(n, length(valid_ports))
+  ports <- sample(valid_ports, n)
+  for (port in ports) {
+    res <- NULL
+
+    # Try to run `.f()` with the random port
+    res <- tryCatch(
+      startup(host = host, port = port),
+      error_timeout = identity,
+      system_command_error = identity,
+      error = function(err) if (identical(port, ports[n])) err
```

Comment:
I'm finding this code hard to reason about. I think part of what makes this code hard for me to understand is that `res` can be used to represent both the return value for `startup()` or the error object that's thrown. It would be easier to understand if those things were kept separate, like this:


```R
for (port in ports) {
  res <- NULL
  err <- NULL
  
  tryCatch({
      res <- startup(host = host, port = port)
    },
    error_timeout = function(e) err <<- e,
    system_command_error = function(e) err <<- e,
    # For other errors, they are probably because the port is already in use.
    # Don't do anything; we'll just continue in the loop.
    error = function(e) NULL
  )
  
  if (!is.null(err)) {
    rlang::cnd_signal(err)
  }
  
  if (!is.null(res)) {
    return(res)
  }
}

rlang::abort(
  "Cannot find an available port. Please try again.",
  class = "chromote_error_no_available_port"
)
```

Reply:
100%. I used your suggestion to separate `res` and `err`. I also simplified the non-generic classes; it's spread over a couple of commits, but in the end I chose `error_stop_port_search` as the class to signal a fatal error that `with_random_port()` should re-throw.